static/src/javascripts/projects/common/modules/commercial/track-ad.js

```diff
@@ -0,0 +1,30 @@
+define([
+    'Promise',
+    'common/utils/mediator'
+], function(Promise, mediator) {
+    return trackAd;
+
+    function trackAd(id) {
+        return new Promise(function (resolve, reject) {
+            var onAdLoaded = function (event) {
+                if (event.slot.getSlotElementId() === id) {
```

Comment:
I would agree that we should track lifecycle with promises, not events. The biggest issue with events is that your listeners must always be registered before the adverts are loaded, which is a kind of coupling across concerns. The other is that each thing that cares about the state of your adverts maintains its own copy of the ad's state.

You could implement a promise-based interface in dfp-api without too much trouble, I reckon:

```javascript
dfpApi.trackAd = memoize((adSlotId) => {
	const advert = adverts[adSlotId];
	const isRendered = advert && advert.isRendered;

	return new Promise((resolve, reject) => {
		if (isRendered) {
			resolve(true);
		} else {
			mediator.once('modules:commercial:dfp:rendered:' + adSlotId, event => {
				const adLoaded = !event.isEmpty;
				resolve(adLoaded);
			});
		}
	});
});
```